components/homme/src/theta-l_kokkos/cxx/LimiterFunctor.hpp

```diff
@@ -0,0 +1,215 @@
+
+#ifndef HOMMEXX_LIMITER_FUNCTOR_HPP
+#define HOMMEXX_LIMITER_FUNCTOR_HPP
+
+#include "Types.hpp"
+#include "Elements.hpp"
+#include "ColumnOps.hpp"
+#include "FunctorsBuffersManager.hpp"
+#include "HybridVCoord.hpp"
+#include "KernelVariables.hpp"
+#include "ReferenceElement.hpp"
+#include "SimulationParams.hpp"
+#include "kokkos_utils.hpp"
+
+#include "utilities/SubviewUtils.hpp"
+#include "utilities/ViewUtils.hpp"
+
+#include "profiling.hpp"
+#include "ErrorDefs.hpp"
+
+#include <assert.h>
+
+namespace Homme {
+
+struct LimiterFunctor {
+
+  struct Buffers {
+    static constexpr int num_3d_scalar_mid_buf = 1;
+    ExecViewUnmanaged<Scalar*    [NP][NP][NUM_LEV]  >   buffer1;
+  };
+
+  int                 m_np1;
+  const int           m_num_elems;
+  const bool          m_theta_hydrostatic_mode;
+
+  HybridVCoord        m_hvcoord;
+  ElementsState       m_state;
+  Buffers             m_buffers;
+  ElementsGeometry    m_geometry;
+
+  struct TagDp3dLimiter {};
+
+  // Policies
+#ifndef NDEBUG
+  template<typename Tag>
+  using TeamPolicyType = Kokkos::TeamPolicy<ExecSpace,Kokkos::LaunchBounds<512,1>,Tag>;
+#else
+  template<typename Tag>
+  using TeamPolicyType = Kokkos::TeamPolicy<ExecSpace,Tag>;
+#endif
+
+  TeamPolicyType<TagDp3dLimiter>   m_policy_dp3d_lim;
+
+  TeamUtils<ExecSpace> m_tu;
+
+  LimiterFunctor(const Elements &elements, const HybridVCoord &hvcoord, const SimulationParams& params)
+      : m_num_elems(elements.num_elems())
+      , m_theta_hydrostatic_mode(params.theta_hydrostatic_mode)
+      , m_hvcoord(hvcoord)
+      , m_state(elements.m_state)
+      , m_geometry(elements.m_geometry)
+      , m_policy_dp3d_lim (Homme::get_default_team_policy<ExecSpace,TagDp3dLimiter>(m_num_elems))
+      , m_tu(m_policy_dp3d_lim)
+  {
+    m_np1 = -1;
+  }
+
+  int requested_buffer_size () const {
+    // Ask the buffers manager to allocate enough buffers
+    const int nslots = m_tu.get_num_ws_slots();
+
+    int num_scalar_mid_buf = Buffers::num_3d_scalar_mid_buf;
+
+    return num_scalar_mid_buf  *NP*NP*NUM_LEV  *VECTOR_SIZE*nslots ;
+  }
+
+  void init_buffers (const FunctorsBuffersManager& fbm) {
+    Errors::runtime_check(fbm.allocated_size()>=requested_buffer_size(), "Error! Buffers size not sufficient.\n");
+
+    Scalar* mem = reinterpret_cast<Scalar*>(fbm.get_memory());
+    const int nslots = m_tu.get_num_ws_slots();
+
+    // Midpoints scalars
+    m_buffers.buffer1       = decltype(m_buffers.buffer1)(mem,nslots);
+    mem += m_buffers.buffer1.size();
+
+    assert ((reinterpret_cast<Real*>(mem) - fbm.get_memory())==requested_buffer_size());
+  }
+
+
+  void run (const int& tl)
+  {
+    profiling_resume();
+
+    GPTLstart("caar limiter");
+    m_np1 = tl;
+    Kokkos::parallel_for("caar loop dp3d limiter", m_policy_dp3d_lim, *this);
+    Kokkos::fence();
+    GPTLstop("caar limiter");
+
+    profiling_pause();
+  }
+
+
+//NEED TO THROW IN SOME MESSAGE
```

Comment:
will check Andrew's scream PR on diagnostics and add a message to the limiter in the same way.